collectd/collectd.go

```diff
@@ -0,0 +1,216 @@
+package collectd
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/kimor79/gollectd"
+)
+
+// DefaultPort for colletd is 25826
+const DefaultPort = 25826
+
+var (
+	// ErrBindAddressRequired is returned when starting the Server
+	// without a TCP or UDP listening address.
+	ErrBindAddressRequired = errors.New("bind address required")
+
+	// ErrDatabaseNotSpecified retuned when no database was specified in the config file
+	ErrDatabaseNotSpecified = errors.New("database was not specified in config")
+
+	// ErrCouldNotParseTypesDBFile returned when unable to parse the typesDBfile passed in
+	ErrCouldNotParseTypesDBFile = errors.New("could not parse typesDBFile")
+
+	// ErrServerClosed return when closing an already closed graphite server.
+	ErrServerClosed = errors.New("server already closed")
+
+	// ErrResolveUDPAddr returned when we are unable to resolve a udp address
+	ErrResolveUDPAddr = errors.New("unable to resolve UDP address")
+
+	// ErrListenUDP returned when we are unable to resolve a udp address
+	ErrListenUDP = errors.New("unable to listen on UDP")
+)
+
+// SeriesWriter defines the interface for the destination of the data.
+type SeriesWriter interface {
+	WriteSeries(database, retentionPolicy, name string, tags map[string]string, timestamp time.Time, values map[string]interface{}) error
+}
+
+type Server struct {
+	mu sync.Mutex
+	wg sync.WaitGroup
+
+	conn *net.UDPConn
+
+	writer      SeriesWriter
+	Database    string
+	typesdb     gollectd.Types
+	typesdbpath string
+}
+
+func NewServer(w SeriesWriter, typesDBPath string) *Server {
+	s := Server{
+		writer:      w,
+		typesdbpath: typesDBPath,
+		typesdb:     make(gollectd.Types),
+	}
+
+	return &s
+}
+
+func ListenAndServe(s *Server, iface string) error {
+	if iface == "" { // Make sure we have an address
+		return ErrBindAddressRequired
+	} else if s.Database == "" { // Make sure they have a database
+		return ErrDatabaseNotSpecified
+	}
+
+	addr, err := net.ResolveUDPAddr("udp", iface)
```

Comment:
As I understand it, the contribution @davecheney brought to the errors discussion is that treating errors as values (or concrete types) is bad in general (be it having global error variables or parsing messages out of error strings). The better approach is treating errors as behaviors -- that is to say, interfaces -- that have methods for additional information you want to pull out of them. For example `interface { error; IsTemporary() bool }` would be good, but `struct { errmsg string; temporary bool }` would be bad